Add CI action to push images to Docker Hub

DanielCosme opened the pull request:
This PR creates a Github Action that builds and push docker images to Docker Hub. For now the trigger to this workflow will be manual and will be tagged only with `latest`. We expect to do a second pass on this workflow so that we have multiple tags that relate to different versions of AM. We will do this once we have a better understanding on how a dockerised production grade deployment of AM looks like.

Also, the Dockerfile has been reworked a little bit to make the resulting images smaller in the registry. We did this by only using dependencies that are strictly needed for all components of AM. Before all 3 components shared dependencies that where not really needed.

I expect to do an additional pass on this file in the future, once I have a better understanding on how the dependencies work. As for the `archivematica-tests` target I left it as it was originally.

DanielCosme committed:
Add GHA to build and push images to docker hub

DanielCosme committed:
Make Docker images smaller

replaceafill commented on .github/workflows/push-images.yml:
```diff
@@ -0,0 +1,46 @@
+on: workflow_dispatch
+jobs:
+  build:
+    name: "Builds images and push them to Docker Hub"
```
```suggestion
    name: "Build and push images"
```

DanielCosme replied:
Done.

replaceafill commented on .github/workflows/push-images.yml:
```diff
@@ -0,0 +1,46 @@
+        uses: "actions/checkout@v4"
+      - name: "Set up buildx"
+        uses: "docker/setup-buildx-action@v3"
+        with:
+          version: latest
+          driver-opts: image=moby/buildkit:v0.13.0
```
I noticed the Storage Service [related change](https://github.com/artefactual/archivematica-storage-service/blob/dd7b6b288173a1d8117993479fd13c6cbc8ea24a/.github/workflows/push-images.yml#L10-L11) doesn't include this. Why is it necessary for the Archivematica images?

DanielCosme replied:
Because the environment in which this is executed uses by default buildkit `v0.12` and we need `v0.13`+ in order for the multitarget feature to work. So we have to directly specify the version.

replaceafill commented on .github/workflows/push-images.yml:
```diff
@@ -0,0 +1,46 @@
+          load: true
+          file: ./hack/Dockerfile
+          target: "archivematica-mcp-server"
+          tags: artefactual/archivematica-mcp-server:latest
```
Could we add a newline at the end of the file?

DanielCosme replied:
Done.

replaceafill commented on hack/Dockerfile:
```diff
@@ -150,6 +150,35 @@ ARG USER_ID=1000
+		coreutils \
+		clamav \
+		libldap2-dev \
+		libsasl2-dev \
```
Could these be removed since they're already installed in the `base-builder`?

DanielCosme replied:
Done

replaceafill commented on hack/Dockerfile:
```diff
@@ -150,6 +150,35 @@ ARG USER_ID=1000
+	&& apt-get update \
+	&& apt-get install -y --no-install-recommends \
+		coreutils \
+		clamav \
```
I might be wrong, but I think the `clamav` package and the:

```
# Download ClamAV virus signatures
RUN freshclam --quiet
```

instruction below are only necessary in the MCPClient stage.

DanielCosme replied:
I'll try it.

DanielCosme replied:
Done. Nothing broke in the tests.

replaceafill commented on hack/Dockerfile:
```diff
@@ -150,6 +150,35 @@ ARG USER_ID=1000
+RUN set -ex \
+	&& apt-get update \
+	&& apt-get install -y --no-install-recommends \
+		coreutils \
```
I noticed the [ansible-archivematica-src vars](https://github.com/artefactual-labs/ansible-archivematica-src/blob/62356e1923fd7c1293391244f1befe42d737ccf8/vars/Ubuntu-22.yml#L43) set this as a dashboard dependency. Should it be moved there?

DanielCosme replied:
I remember. I decided to leave `coreutils` here just because I don't understand the components and dependencies enough. coreutils contains (sometimes) essential packages for the OS. And since AM is so reliant on OS dependencies I left it in the `base-builder` step. Should I remove it?

replaceafill replied:
I think you can try moving `coreutils` to the dashboard stage and let the Acceptance Test workflow to tell you if anything breaks.

DanielCosme replied:
`coreutils` removed. nothing broke :)

replaceafill commented on hack/Dockerfile:
```diff
@@ -283,7 +296,75 @@ ENTRYPOINT ["pyenv", "exec", "python3", "-m", "gunicorn", "--config=/src/src/das
 # -----------------------------------------------------------------------------
 
-FROM base AS archivematica-tests
+FROM base-builder as archivematica-tests
```
I think the `archivematica-tests` stage can be much more simple. It's needed for running `tox` and building its virtual environments, so it only needs Python, the common development libraries `base-builder` already has and from a quick local test based on the `qa/1.x` branch the following dependencies:

* `gcc` for building C dependencies for the virtual environment.
* `media-types`, `p7zip-full`, `rsync` and `unar` for being able to run the Storage Service tests from the `hack/Makefile`.

DanielCosme replied:
Yeah, it works, thank you. Done.

replaceafill commented on hack/Dockerfile:
```diff
@@ -283,7 +296,75 @@ ENTRYPOINT ["pyenv", "exec", "python3", "-m", "gunicorn", "--config=/src/src/das
+		uuid \
+	&& rm -rf /var/lib/apt/lists/* /var/cache/apt/*
+
+RUN set -ex \
+	&& groupadd --gid ${GROUP_ID} --system archivematica \
+	&& useradd --uid ${USER_ID} --gid ${GROUP_ID} --home-dir /var/archivematica --system archivematica \
+	&& mkdir -p /var/archivematica/sharedDirectory \
+	&& chown -R archivematica:archivematica /var/archivematica
+
+# Download ClamAV virus signatures
+RUN freshclam --quiet
+
+USER archivematica
+
+COPY --chown=${USER_ID}:${GROUP_ID} --from=pyenv-builder --link ${PYENV_DIR} ${PYENV_DIR}
+COPY --chown=${USER_ID}:${GROUP_ID} --link . /src
```
Not sure if it makes sense, but I think if you end up moving all the OS dependencies out of `base` to their corresponding stages, this block might be inherited from the `base` stage instead.

DanielCosme replied:
So what I think you are saying is that this block is not really needed because it already happens in `base`. So by having this block here is just redundant, doing the same work again, right?

replaceafill replied:
That's correct.

DanielCosme replied:
Done.

replaceafill reviewed (commented):
@DanielCosme great progress! The new workflow for publishing images looks good to me. I haven't tested the `Dockerfile` in the PR yet, but I left a few questions and suggestions.

sevein commented on .github/workflows/push-images.yml:
```diff
@@ -0,0 +1,46 @@
+          file: ./hack/Dockerfile
+          target: "archivematica-dashboard"
+          tags: artefactual/archivematica-dashboard:latest
+      - name: "Build and Push MCP-client"
```
```suggestion
      - name: "Build and push MCPClient"
```

DanielCosme replied:
Done.

sevein commented on .github/workflows/push-images.yml:
```diff
@@ -0,0 +1,46 @@
+          file: ./hack/Dockerfile
+          target: "archivematica-mcp-client"
+          tags: artefactual/archivematica-mcp-client:latest
+      - name: "Build and Push MCP-server"
```
```suggestion
      - name: "Build and push MCPServer"
```

DanielCosme replied:
Done.

sevein commented on hack/Dockerfile:
```diff
@@ -283,7 +296,75 @@ ENTRYPOINT ["pyenv", "exec", "python3", "-m", "gunicorn", "--config=/src/src/das
-FROM base AS archivematica-tests
+FROM base-builder as archivematica-tests
+
+ARG USER_ID=1000
+ARG GROUP_ID=1000
+ARG PYENV_DIR=/pyenv
```
According to the docs, we don't need to redefine the default again when pulling the argument, e.g.:

```suggestion
ARG USER_ID
ARG GROUP_ID
ARG PYENV_DIR
```

DanielCosme replied:
Thanks!

DanielCosme committed:
WIP: feedback from code review

DanielCosme committed:
WIP: fix typos in GH workflow

DanielCosme committed:
WIP: trigger CI

DanielCosme committed:
Remove coreutils comment

DanielCosme committed:
Revert back Github action trigger to manual

DanielCosme commented:
@replaceafill I think I addressed all feedback. All tests are passing, if you have any further feedback, let me know.

sevein committed:
Introduce JSONDataEncoder for Gearman

sevein committed:
Use JSONDataEncoder in MCP task backend

sevein committed:
Use JSONDataEncoder in RPC server

sevein committed:
Remove use of pickle in Access.target

DanielCosme committed:
Add SSH package to mcp-client

sevein committed:
Remove dead code in MCPServer

sevein committed:
Remove unused file filter

sevein committed:
Use json.loads in OutputClientScriptJob

sevein committed:
Use orjson in JSONDataEncoder

orjson is significantly faster and comes with native serialization of some
types like datetime or UUID.

sevein committed:
Use JSONDataEncoder to encode task createdDate

This commit modifies the MCPServer to encode the `createdDate` of tasks using
`datetime` which is natively supported by our JSONDataEncoder. The conversion
to the METS-compliant format now occurs during the decoding process on the
client side. This change ensures consistent use of RFC 3339 format across all
communications.

replaceafill committed:
Remove comment from JHOVE validation migration

replaceafill committed:
Fix default thumbnail normalization command

This adds a data migration for replacing the existing command with a
fixed version of its Python script.

DanielCosme committed:
WIP: trim build time dependencies from final images

replaceafill committed:
Extend elasticSearchFunctions test coverage

replaceafill committed:
Fix SIP arrangement from ArchivesSpace pane

replaceafill committed:
Remove references to Binder

replaceafill committed:
Fix failure reports index

replaceafill committed:
Upgrade Python requirements

replaceafill committed:
Update submodules

replaceafill committed:
Update archivematica-storage-service submodule

replaceafill committed:
Bump version to 1.17.0

replaceafill committed:
Upgrade Python requirements

sevein committed:
Add SYS_NICE capability to MySQL container

MySQL uses mbind to bind processes and threads to CPUs which is not allowed by
the default configuration of Docker's seccomp profile. This causes MySQL to log
the following error repeatedly:

    mbind: Operation not permitted

Adding the `SYS_NICE` capability to the container forces Docker to adjust the
default seccomp profile to include the necessary permissions. For more details,
visit https://github.com/docker-library/mysql/issues/303.

It looks like MySQL fixed the problem in [v8.0.13] but Percona doesn't seem to
have received this update.

[v8.0.13]: https://dev.mysql.com/doc/relnotes/mysql/8.0/en/news-8-0-13.html

replaceafill committed:
Use Ruff for Python linting and formatting

This replaces the black, pyupgrade, reorder-python-imports and flake8           
hooks in pre-commit with ruff.                                                  

It also upgrades the django-upgrade and markdownlint-cli hooks.

replaceafill committed:
Allow triggering the test CI workflows manually

replaceafill commented:
Closing this after we decided to approach it in two separate pull requests:
* Add a GitHub workflow to publish images to Docker Hub. See https://github.com/artefactual/archivematica/pull/2070
* Make the Docker image sizes as small as possible
